### PR TITLE
spanconfigsqltranslator: emit all SystemTarget span configs when required

### PR DESCRIPTION
**Pull request by adityamaru**
This change teaches the SQLTranslator to emit SpanConfigurations
corresponding to `spanconfig.SystemTargets`. Today, these SpanConfigurations
only contain ProtectionPolicies, corresponding to protected timestamp
records that may be written by the host tenant to protect its cluster,
a secondary tenant to protect its cluster, or the host tenant to protect
a secondary tenant.

The SystemTarget span configurations will be applied to a SystemSpanConfig
store that will be introduced in a follow up PR.

Informs: https://github.com/cockroachdb/cockroach/issues/73727

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
First two commits from https://github.com/cockroachdb/cockroach/pull/76414, will rebase once it merges.

**Review by arulajmani — COMMENTED** [review]
I mostly have nitty comments! Good stuff :100: 

> Reviewed 8 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-:-Mw2n3hyAR1-5zfGjojH:b-u8mbal)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @irfansharif)

___
*[pkg/roachpb/span_config.go, line 113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2j51bBwJzDIETOEiN:-Mw2j51bBwJzDIETOEiO:bi6t2h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/roachpb/span_config.go#L113)):*
> ```Go
> // TestingSystemTargetDefaultSpanConfig exports the default span config that
> // applies to spanconfig.SystemTargets for testing purposes.
> func TestingSystemTargetDefaultSpanConfig() SpanConfig {
> ```

Should we call this `TestingDefaultSystemSpanConfiguration`? 
___
*[pkg/spanconfig/spanconfig.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2jP_Z-1lxBfU5XfQi:-Mw2jP_Z-1lxBfU5XfQj:b-kfanfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfig.go#L116)):*
> ```Go
> 	// configuration. Translate also accounts for and negotiates subzone spans.
> 	Translate(ctx context.Context, ids descpb.IDs,
> 		shouldGenerateSystemTargetRecords bool) ([]Record, hlc.Timestamp, error)
> ```

Similar to my comment above, should we call this `generateSystemSpanConfigurations`? 
___
*[pkg/spanconfig/spanconfigreconciler/reconciler.go, line 382 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2k4PkDhFEo1ksjZ7V:-Mw2k4PkDhFEo1ksjZ7W:b6e4rib) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfigreconciler/reconciler.go#L382)):*
> ```Go
> 					allIDs = append(allIDs, update.GetDescriptorUpdate().ID)
> 				} else if update.IsProtectedTimestampUpdate() {
> 					shouldGenerateSystemTargetRecords = true
> ```

Should we revert this until we teach + test system span configurations to the reconciler? Wiring this up right now, without the KV changes, might cause issues. 
___
*[pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2lK-_Agwsfj7TYo68:-Mw2lK-_Agwsfj7TYo69:b39sf9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go#L157)):*
> ```Go
> // generateSystemTargetRecords is responsible for generating all the SpanConfigs
> // that apply to spanconfig.SystemTargets.
> func (s *SQLTranslator) generateSystemTargetRecords(
> ```

s/generateSystemTargetRecords/generateSystemSpanConfigRecords/?
___
*[pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2mqc_4OPkmgddNd-j:-Mw2mqc_4OPkmgddNd-k:bbta5l3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go#L187)):*
> ```Go
> 
> 	// Aggregate tenant target protections.
> 	tenantProtections := ptsStateReader.getProtectionPoliciesForTenants()
> ```

Centralizing the discussion we've had in a couple of places, do you prefer this approach over returning ptpb.Target's from the ptsStateReader and writing conversions from a ptpb.Target to a spanconfig.Target (and vice versa)? 
___
*[pkg/spanconfig/spanconfigtestutils/utils.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2mboWAzsQIljb8AQS:-Mw2mboWAzsQIljb8AQT:bhnek9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfigtestutils/utils.go#L277)):*
> ```Go
> // diffs the given config against the default config that applies to
> // spanconfig.SystemTargets, and returns a string for the mismatched fields.
> func PrintSystemTargetSpanConfigDiffedAgainstDefault(conf roachpb.SpanConfig) string {
> ```

s/PrintSystemTargetSpanConfigDiffedAgainstDefault/PrintSystemSpanConfigDiffedAgainstDefault/g
___
*[pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2khVr2qlVgmuOmqiQ:-Mw2khVs2hOICL5M_xhq:bhxw3le) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go#L174)):*
> ```Go
> 						output.WriteString(fmt.Sprintf("%-42s %s\n", record.Target.GetSpan(),
> 							spanconfigtestutils.PrintSpanConfigDiffedAgainstDefaults(record.Config)))
> 					} else {
> ```

nit: might be worth using a switch statement which explicitly checks for `IsSpanTarget` and `IsSystemTarget` and panics in the default case.

**Comment by adityamaru** [comment]
Going to wait for #76721 before I rebase this PR.

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @irfansharif)

___
*[pkg/roachpb/span_config.go, line 113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2j51bBwJzDIETOEiN:-MwRfLLP05F4Z74rL_sZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/roachpb/span_config.go#L113)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we call this `TestingDefaultSystemSpanConfiguration`? 
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfig.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2jP_Z-1lxBfU5XfQi:-MwRfcE7ABEU5ThOkuuS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfig.go#L116)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Similar to my comment above, should we call this `generateSystemSpanConfigurations`? 
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigreconciler/reconciler.go, line 382 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2k4PkDhFEo1ksjZ7V:-MwRfdmF1PfJgUZRVfCy:b-hqaudb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfigreconciler/reconciler.go#L382)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we revert this until we teach + test system span configurations to the reconciler? Wiring this up right now, without the KV changes, might cause issues. 
</blockquote></details>

Sounds good, added a todo.
___
*[pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2lK-_Agwsfj7TYo68:-MwRg52v1kVFkPR31bH5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go#L157)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

s/generateSystemTargetRecords/generateSystemSpanConfigRecords/?
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2mqc_4OPkmgddNd-j:-MwRgbdN8kgPYI3t8XTJ:bc9jadp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go#L187)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Centralizing the discussion we've had in a couple of places, do you prefer this approach over returning ptpb.Target's from the ptsStateReader and writing conversions from a ptpb.Target to a spanconfig.Target (and vice versa)? 
</blockquote></details>

After the discussions/refactor of the methods in your other PRs, I'm happy with where this is right now. If we think there's some cleanup to be done in the future we can revisit.
___
*[pkg/spanconfig/spanconfigtestutils/utils.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2mboWAzsQIljb8AQS:-MwRgbEa4wVZSLcrEHRd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/spanconfig/spanconfigtestutils/utils.go#L277)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

s/PrintSystemTargetSpanConfigDiffedAgainstDefault/PrintSystemSpanConfigDiffedAgainstDefault/g
</blockquote></details>

Done.
___
*[pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-Mw2khVr2qlVgmuOmqiQ:-MwRjYYKAG8yVEtJsKxw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d578aa5d95a1ca820e84060cb33ded538182717f/pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go#L174)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: might be worth using a switch statement which explicitly checks for `IsSpanTarget` and `IsSystemTarget` and panics in the default case. 
</blockquote></details>

Done.

**Comment by adityamaru** [comment]
@arulajmani rebased, and addressed all your comments. Should be RFAL

**Review by arulajmani — APPROVED** [review]
:lgtm: once we introduce a new directive to translate system span configurations 

> Reviewed 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-:-MwUWWIL2BMN-4RDnKPv:b-l2we1y)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @irfansharif)

___
*[pkg/spanconfig/spanconfigtestutils/utils.go, line 280 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-MwUUwLIDE7oTDSWPdOV:-MwUUwLIDE7oTDSWPdOW:bi7jxff) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb2f78da22d48721705bbd221ea6a9a897ce319/pkg/spanconfig/spanconfigtestutils/utils.go#L280)):*
> ```Go
> 
> // PrintSystemSpanConfigDiffedAgainstDefault is a helper function that
> // diffs the given config against the default config that applies to
> ```

nit: "the default system span config" 
___
*[pkg/spanconfig/spanconfigtestutils/utils.go, line 284 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-MwUV0289ui_gtoyPlnE:-MwUV0289ui_gtoyPlnF:bpf34l0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb2f78da22d48721705bbd221ea6a9a897ce319/pkg/spanconfig/spanconfigtestutils/utils.go#L284)):*
> ```Go
> func PrintSystemSpanConfigDiffedAgainstDefault(conf roachpb.SpanConfig) string {
> 	if conf.Equal(roachpb.TestingDefaultSystemSpanConfiguration()) {
> 		return "system target default"
> ```

nit: "default system span config"
___
*[pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76606#-MwUViRP9Iah9oe2OTTG:-MwUViRQ0b7XO_PEjAWI:bfte1ev) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb2f78da22d48721705bbd221ea6a9a897ce319/pkg/ccl/spanconfigccl/spanconfigsqltranslatorccl/datadriven_test.go#L163)):*
> ```Go
> 				sqlTranslator := tenant.SpanConfigSQLTranslator().(spanconfig.SQLTranslator)
> 				records, _, err := sqlTranslator.Translate(ctx, descpb.IDs{objID},
> 					true /* generateSystemSpanConfigurations */)
> ```

Should we create a new directive to just generate system span configurations? As is, it's slightly confusing that these things are getting generated when translating a table or a database.

**Comment by adityamaru** [comment]
Addressed all your comments, TFTR!

**Comment by adityamaru** [comment]
bors r=arulajmani

**Comment by adityamaru** [comment]
Merge conflict, rebasing and trying again.

**Comment by adityamaru** [comment]
bors r=arulajmani

**Comment by adityamaru** [comment]
Unrelated flake, lets try this again.

bors r=arulajmani

**Pull request merged**
